src/backend/session.rs

```diff
@@ -0,0 +1,193 @@
+use std::{
+    collections::HashMap,
+    fmt::{Debug, Display, Formatter},
+    sync::{Arc, Mutex},
+};
+
+use async_trait::async_trait;
+use itertools::Itertools;
+use zbus::{
+    object_server::SignalEmitter,
+    zvariant::{OwnedObjectPath, SerializeDict, Type},
+};
+
+use crate::{desktop::HandleToken, error::PortalError};
+
+pub(crate) struct Session {
+    path: OwnedObjectPath,
+    manager: Arc<Mutex<SessionManager>>,
+    monitor: Option<Arc<dyn SessionImpl>>,
+}
+
+impl Session {
+    pub(crate) fn new(
+        path: OwnedObjectPath,
+        manager: Arc<Mutex<SessionManager>>,
+        monitor: Option<Arc<dyn SessionImpl>>,
+    ) -> Self {
+        Self {
+            path,
+            manager,
+            monitor,
+        }
+    }
+
+    pub fn handle_token(&self) -> HandleToken {
+        HandleToken::try_from(&self.path).unwrap()
+    }
+
+    pub async fn serve(&self, cnx: zbus::Connection) -> zbus::Result<bool> {
+        let interface = SessionInterface::new(
+            self.path.clone(),
+            Arc::clone(&self.manager),
+            self.monitor.clone(),
+        );
+        cnx.object_server().at(&self.path, interface).await
+    }
+}
+
+impl PartialEq for Session {
+    fn eq(&self, other: &Self) -> bool {
+        self.path == other.path
+    }
+}
+
+struct SessionInterface {
+    path: OwnedObjectPath,
+    manager: Arc<Mutex<SessionManager>>,
+    monitor: Option<Arc<dyn SessionImpl>>,
+}
+
+impl SessionInterface {
+    fn new(
+        path: OwnedObjectPath,
+        manager: Arc<Mutex<SessionManager>>,
+        monitor: Option<Arc<dyn SessionImpl>>,
+    ) -> Self {
+        Self {
+            path,
+            manager,
+            monitor,
+        }
+    }
+}
+
+#[zbus::interface(name = "org.freedesktop.impl.portal.Session")]
+impl SessionInterface {
+    #[zbus(property(emits_changed_signal = "const"), name = "version")]
+    fn version(&self) -> u32 {
+        1
+    }
+
+    async fn close(
+        &self,
+        #[zbus(object_server)] server: &zbus::ObjectServer,
+    ) -> zbus::fdo::Result<()> {
+        #[cfg(feature = "tracing")]
+        tracing::debug!("SessionInterface::Close {}", self.path.as_str());
+        let handle_token = HandleToken::try_from(&self.path).unwrap();
+        {
+            let mut manager = self.manager.lock().unwrap();
+            let _ = manager.remove(&handle_token);
+        }
+        if let Some(monitor) = &self.monitor {
+            let _ = monitor.session_closed(handle_token).await;
```

Comment:
I guess we said, no session_closed emitted when you Close is called?

Reply:
Instead, it should be called if Session is dropped

Reply:
Perhaps this is a bit confusing, but calling `monitor.session_closed` isn't emitting the `Closed` signal on the session interface, it merely notifies the backend implementer that a session was closed, in case they want to do some cleanup on their side.

Reply:
ah then, please make sure it is properly documented